src/System.Private.CoreLib/src/System/Text/Utf8Span.Conversion.cs

```diff
@@ -0,0 +1,321 @@
+            // we can perform transcoding using an optimized code path that skips all safety checks.
+            // We should also consider skipping the two-pass if possible.
+
+            fixed (byte* pbUtf8 = &DangerousGetMutableReference())
+            {
+                byte* pbUtf8Invalid = Utf8Utility.GetPointerToFirstInvalidByte(pbUtf8, this.Length, out int utf16CodeUnitCountAdjustment, out _);
+                Debug.Assert(pbUtf8Invalid == pbUtf8 + this.Length, "Invalid UTF-8 data seen in buffer.");
+
+                char[] asUtf16 = new char[this.Length + utf16CodeUnitCountAdjustment];
+                fixed (byte* pbUtf16 = &asUtf16.GetRawSzArrayData())
+                {
+                    OperationStatus status = Utf8Utility.TranscodeToUtf16(pbUtf8, this.Length, (char*)pbUtf16, asUtf16.Length, out byte* pbUtf8End, out char* pchUtf16End);
+                    Debug.Assert(status == OperationStatus.Done, "The buffer changed out from under us unexpectedly?");
+                    Debug.Assert(pbUtf8End == pbUtf8 + this.Length, "The buffer changed out from under us unexpectedly?");
+                    Debug.Assert(pchUtf16End == ((char*)pbUtf16) + asUtf16.Length, "The buffer changed out from under us unexpectedly?");
+
+                    return asUtf16;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Converts this <see cref="Utf8Span"/> instance to its UTF-16 equivalent, writing the result into
+        /// the buffer <paramref name="destination"/>.
+        /// </summary>
+        /// <returns>
+        /// The number of bytes written to <paramref name="destination"/>, or -1 if <paramref name="destination"/>
+        /// is not large enough to hold the result of the transcoding operation.
+        /// </returns>
+        public int ToChars(Span<char> destination)
+        {
+            OperationStatus status = Utf8.ToUtf16(Bytes, destination, out int _, out int charsWritten, replaceInvalidSequences: false, isFinalBlock: true);
+
+            Debug.Assert(status == OperationStatus.Done || status == OperationStatus.DestinationTooSmall, "Utf8Spans shouldn't contain ill-formed UTF-8 data.");
+
+            if (status != OperationStatus.Done)
+            {
+                charsWritten = -1; // "destination too small"
+            }
+
+            return charsWritten;
+        }
+
+        /// <summary>
+        /// Returns a new <see cref="Utf8String"/> instance which represents this <see cref="Utf8Span"/> instance
+        /// converted to lowercase using <paramref name="culture"/>.
+        /// </summary>
+        /// <remarks>
+        /// The original <see cref="Utf8Span"/> is left unchanged by this operation. Note that the returned
+        /// <see cref="Utf8String"/> instance may be longer or shorter (in terms of UTF-8 byte count) than the
+        /// input <see cref="Utf8Span"/>.
+        /// </remarks>
+        public Utf8String ToLower(CultureInfo culture)
```

Comment:
The APIs for non-ordinal globalization are slow and complex. Do we really need globalization support for UTF8 strings? Do we have enough evidence that there is enough benefit in creating UTF8 globalization clone? Can we get away by supporting just ordinal comparisons, etc. for UTF8 strings?

Reply:
At minimum we'll need `Ordinal` and `OrdinalIgnoreCase`. This is the reason I've been making targeted changes to `char` and `CharUnicodeInfo` - getting things ready for us carrying additional Unicode data within System.Runtime.CoreLib and not needing to p/invoke NLS.

Culture-aware globalization (even for the "invariant" culture) is a stickier issue. For web services I don't think they'll be necessary. But for applications which generate UI - whether they're running on the client or as part of a server web rendering framework - I think the lack of such APIs will ultimately harm the usability of the types. But honestly this is just a first-level guess and I don't have good evidence to support the theory that these scenarios will be important moving forward.